Implement more natural human player input.

lucasb-eyer opened the pull request:
The comment on top should describe both the problem of the current approach, as well as the new proposed approach, so I won't repeat it here. Code is extensively commented as otherwise it looks needlessly over-engineered, but it isn't.

The magic numbers used (`0.2` and `0.1`) are what I remember using in past games, but could be tweaked. For example, my linux uses `0.66` and `0.025` (check yours using `xset q | grep 'auto repeat delay'`) but I felt it was too delayed for a game.

Getting user input feel "natural" is actually nontrivial (I know from experience) and super-simple code almost never feels right. I know it wasn't a priority for this project, but it felt frustrating play-testing it and losing to `SimpleAgent` because of input issues, not skill. With this patch, I'm still losing regularly, but at least now it feels like I'm losing because of my poor skill :smile: I also play-tested it in multiplayer with my wife and it works nice.

Note that this PR comments-out the pause action since it crashes the game (as noted marked in a comment before my changes already) and we hit it several times by accident during play. I can undo that and leave it in if you prefer.

cinjon commented on pommerman/agents/player_agent.py:
```diff
@@ -1,13 +1,70 @@
+class Keystate:
+    def __init__(self):
+        self.t0 = time()
+        self.tlast = None
```
Please use explicit names, e.g. self.tlast --> self.last_time or self.prev_time

cinjon commented on pommerman/agents/player_agent.py:
```diff
@@ -1,13 +1,70 @@
+# NOTE:
```
I like all of this documentation. Please wrap it in a comment string instead of "#"

cinjon commented on pommerman/agents/player_agent.py:
```diff
@@ -16,7 +73,7 @@ def __init__(self, agent, control):
                 key.RIGHT: 4,
                 key.SPACE: 5,
-                key.M: 6 # In Pommerman, this will freeze the game.
+                # key.M: 6 # In Pommerman, this will freeze the game.
```
I saw your comment that this crashes the game, however I was not having that issue before your last PR. Can you describe this more? Is there any other behavior that occurs?

lucasb-eyer replied:
This was definitely in before, [see for example here](https://github.com/MultiAgentLearning/playground/blob/373fbd2850231b7de6bf2c7315bd6b12c2b46d09/pommerman/utility.py#L46), which is a commit from before I even knew about this project.

Actually, this might be me misunderstanding "freeze" as "crash" where it originally just means "pause" and it worked well all along. I'll put them back in while addressing your other comments.

lucasb-eyer replied:
Actually, no, it really is a problem, I get this:

```
Traceback (most recent call last):
  File "/home/beyer/academic/pommerman/venv/bin/pom_battle", line 11, in <module>
    load_entry_point('pommerman', 'console_scripts', 'pom_battle')()
  File "/home/beyer/academic/pommerman/playground/pommerman/cli/run_battle.py", line 118, in main
    run(args)
  File "/home/beyer/academic/pommerman/playground/pommerman/cli/run_battle.py", line 80, in run
    infos.append(_run(seed, record_pngs_dir_, record_json_dir_))
  File "/home/beyer/academic/pommerman/playground/pommerman/cli/run_battle.py", line 60, in _run
    actions = env.act(obs)
  File "/home/beyer/academic/pommerman/playground/pommerman/envs/v0.py", line 115, in act
    return self.model.act(agents, obs, self.action_space)
  File "/home/beyer/academic/pommerman/playground/pommerman/envs/utility.py", line 420, in act
    ret.append(act_ex_communication(agent))
  File "/home/beyer/academic/pommerman/playground/pommerman/envs/utility.py", line 396, in act_ex_communication
    time.sleep(300)
NameError: name 'time' is not defined
```

It's just missing an `import time` in `envs/utility.py`, I can fix that in another PR since it has nothing to do with this one. It does look like a weird feature, though, pausing for 5min without possibility to unpause :laughing:

cinjon replied:
Oh, that must have happened in all the refactoring. Thanks for catching it.

Regarding the pausing, it's actually been pretty darn helpful for me for both testing the game world and for verifying my agents' actions with the game state. We now have more machinery in place to do that (e.g. the game state files), so might not be as vital. 

That being said, adjusting it so that the game continues if you press it again would be great 🥇

cinjon commented on pommerman/agents/player_agent.py:
```diff
@@ -1,13 +1,70 @@
+        self.tlast = None
+        self.fired = False
+
+    def fires(self):
```
i think this should be "should_fire"

cinjon replied:
or actually maybe "should_repeat"

cinjon commented on pommerman/agents/player_agent.py:
```diff
@@ -25,20 +82,40 @@ def __init__(self, agent, control):
+        act = 0
+        if self._q:  # Work off the keys that are queued.
+            act = self._key2act.get(self._q.pop(0), 0)
+        # print(act)
```
remove print

cinjon commented on pommerman/agents/player_agent.py:
```diff
@@ -25,20 +82,40 @@ def __init__(self, agent, control):
+
+        act = 0
+        if self._q:  # Work off the keys that are queued.
+            act = self._key2act.get(self._q.pop(0), 0)
```
when would this use the default "0"?

lucasb-eyer replied:
Legacy; you're right, it can't in the current implementation, so I'll change it to `[]`.

cinjon commented on pommerman/agents/player_agent.py:
```diff
@@ -25,20 +82,40 @@ def __init__(self, agent, control):
     def on_key_release(self, k, mod):
-        self._key_input['curr'] = 0
+        # It could not be in dict if it was pressed when focus was outside the
+        # window but released when the window had focus.
```
I don't understand this comment. Can you elaborate please?

lucasb-eyer replied:
I was describing how one can get a `key_release` event without having gotten a corresponding `key_press` event before. But this comment is legacy again, in the current incarnation, the check should be there in all cases! Changing it.

cinjon commented:
Left some comments in the PR. Thanks for this; I really appreciate the detailed comments.

lucasb-eyer commented:
Thanks for the comments! I addressed them all and rebased onto latest master, but I can't test it yet due [this](https://github.com/MultiAgentLearning/playground/pull/40#issuecomment-376984803), so I'll hold off pushing the changes until then.

cinjon commented:
Addressed #48 --> ping to test this.

lucasb-eyer commented:
OK, rebased on master, solved conflicts, and addressed all your comments.

cinjon commented:
lgtm!
